File: Leetcode/C++/39. Combination Sum/solution.cpp
// https://leetcode.com/problems/combination-sum/

#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> result;
    vector<int> numbers;
    int targetSum, size;
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        targetSum = target;
        size = candidates.size();
        numbers = candidates;

        vector<int> comb;
        dfs(comb, 0, 0);
        return result;
    }

    void dfs(vector<int> &comb, int total, int start)
    {
        if (total == targetSum)
        {
            result.push_back(comb);
            return;
        }
        else if (total > targetSum) // Backtracking
        {
            return;
        }

        for (int index = start; index < size; index++)
        {
            // Search current index number after search to avoid affecting search results pop element
            comb.push_back(numbers[index]);
            dfs(comb, total + numbers[index], index);
            comb.pop_back();
        }
    }
};

File: Leetcode/Typescript/39. Combination Sum/solution.ts
// https://leetcode.com/problems/combination-sum/description/

function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = [];

  function dfs(prev: number[], index: number, total: number) {
    if (total === target) {
      result.push(prev.slice());
      return;
    }

    for (let idx = index; idx < candidates.length; idx++) {
      total += candidates[idx];
      prev.push(candidates[idx]);
      // Backtracking
      if (total <= target) {
        dfs(prev, idx, total);
      }
      total -= candidates[idx];
      prev.pop();
    }
  }

  dfs([], 0, 0);

  return result;
}
